smallab/dashboard/dashboard.py

```diff
@@ -39,65 +60,139 @@ def record_iteration(self, specification, progress):
         else:
             progress_amount = progress - self.last_progress[specification]
             self.last_progress[specification] = progress
+
         current_time = time.time()
         time_diff = current_time - self.last_update_time[specification]
+
         self.last_update_time[specification] = current_time
         if progress_amount != 0:
             unit_time_diff = time_diff / progress_amount
         else:
             unit_time_diff = time_diff
 
-        self.time_per_iterations.append(unit_time_diff)
+        self.iteration_dataset.setdefault(specification, []).append(unit_time_diff)
 
     def record_completion(self, specification):
         current_time = time.time()
         time_diff = current_time - self.start_time[specification]
-        self.time_per_completion.append(time_diff)
+        self.completion_dataset.setdefault(specification, []).append(time_diff)
         self.last_update_time[specification] = current_time
 
+        # Record completion in save file: contents -> 'completions': list(dict("spec":spec_entry, "times":list(times)))
+        with open(self.save_file, "r") as f:
+            contents = json.load(f)
+            completions = contents.setdefault('completions', [])
+
+        # Add completion time to dataset
+        spec_entry = self.spec_id_to_entry(specification)
+        found = False
+        for rec in completions:
+            if rec['spec'] == spec_entry:
+                rec['times'].append(time_diff)
+                found = True
+                break
+
+        if not found:
+            completions.append({"spec": spec_entry, "times": [time_diff]})
+
+        # Save updated dataset
+        with open(self.save_file, "w") as f:
+            json.dump(contents, f, indent=4)
+
     def compute_time_stats(self, specification_progress, active, registered):
```

Comment:
There's a backwards compatibility case that's not handled here. Some experiments don't record iterations they only record completion. I don't think we need to handle that right now since no projects really use that.

Reply:
Oh okay. This only records completions though, so would that cause an issue?